### DISCUSSION TITLE
is there a way to read a family provider without passing any values?

### DISCUSSION THREAD

**Discussion by imcouri** [rrousselGit/riverpod#861] [Q&A]
Hey, I'm trying to implement pagination for a ChangeNotifierProvider using a FutureProvider which stores its values in the change notifier provider, the problem is that I can't add the new values to the previously created change notifier and it creates a new change notifier which resets the listview builder, is there a way to read the change notifier which is stored in a state notifier without creating a new instance of it and using the same provider instance which already contains the previous values?

**Comment by rrousselGit** [comment]
I don't quite understand how this relates to `family`.

Do you maybe have a _small_ example of the issue?

**Reply by imcouri** [reply]
Hey thank you for your reply, the code is really long and I'm not sure how i can fit it here so I'll try my best to explain it, I use family to pass the userID to the provider which fetches posts posted by a user, and then those posts are stored in a change notifier provider.

Homepage -> Navigator.push to userPage with user ID -> userID passed to the FutureProivder-> fetched posts stored in a ChangeNotifierProvider and displayed

in case the user reaches the end of the page and hasMorePages boolean which was received from the API is true the future provider will be called again to get 20 more posts and store them in the change notifier, which is the reason I need to read the family provider to add the newly fetched posts the list which contains the previously fetched posts

I could be using a state notifier to pass the userID to the user page, but since you can go to other users pages from one user's page there could be 10 user pages stacked on top of each other, and in that case, they will all show the same user since they are sharing the same state provider to get the user's id.

I tried my best to explain this case as best as I could, please let me know if there's something I missed or if there's something that you'd like to know to help you understand my issue.

Thank you.

**Reply by rrousselGit** [reply]
> in case the user reaches the end of the page and hasMorePages boolean which was received from the API is true the future provider will be called again to get 20 more posts and store them in the change notifier, which is the reason I need to read the family provider to add the newly fetched posts the list which contains the previously fetched posts

Could you expand on this? I don't understand what the "which is the reason" part is about

**Reply by imcouri** [reply]
Sure.
Let's say a user has 100 posts, I'll need to fetch 5 times (since I only receive 20 posts per request) to get all of his/her posts and if I only rely on the future provider I will only see the last 20 posts which were fetched from the API. So every time the future provider gets posts, they get stored in the change notifier, therefore after 5 fetches, all 100 posts will be visible because they are all stored in the change notifier provider. That's why I said I need to access the family provider because I need to add the posts fetched by the future provider to the change notifier provider.

**Reply by rrousselGit** [reply]
You don't need a ChangeNotifier to store the pages for that. You could directly read the FutureProvider in your widgets. But what do you mean by needing to access the family? What prevents you from doing so?

**Reply by imcouri** [reply]
Each response from the API contains 20 posts, so page 1 returns 20, page 2 returns 20 as well. For example, imagine a Twitter feed you see 20 posts and when you scroll down 20 more posts are added to the list every time you reach the end. In my case, the future provider only contains those 20 that were fetched when you reached the bottom of the page and the rest of the posts are stored in the change notifier provider(please let me know if that's now how I should riverpod). Now the future provider is a family provider which takes a userID int and a page int to fetch a user's posts, and the change notifier is also a family because if it's a normal provider user pages can't be stacked and every user page is going to show the same user/posts. 

I tried to show it in this image, thanks for your help

![Untitled Diagram drawio](https://user-images.githubusercontent.com/61738823/139533103-0bf18ee9-e37b-4329-9060-3666904e03e0.png)

**Reply by rrousselGit** [reply]
But what difficulty do you have with reading your provider?

**Reply by imcouri** [reply]
I cant use ref or context.read() on a family without assigning a new value to it so i if I want to read the provider I have to do something like this
`ref.read(postListProvider(passedValue);`
which changes the value of the provider

**Reply by rrousselGit** [reply]
What do you mean by "it changes the value of the provider" and "I can't read without assigning a new value"?
You shouldn't have any problem

**Reply by imcouri** [reply]
`final testProvider = StateProvider.family<int,int>((ref,testValue) {
  return testValue;
});
`

let's say i need to read the value of this provider.

if I read it how I would normally read a provider it would be like this:
`var value = ref.read(testProvider).state;`
but in this case I'd get an error
`error: The argument type 'StateProviderFamily<int, int>' can't be assigned to the parameter type 'RootProvider<Object, dynamic>'.`

so i need to also pass a value with it like this:
`var value = ref.read(testProvider(10)).state;`
 
But this way the value of this provider is overwritten and instead of the value it held, I will receive 10 since I passed 10 as the family value, I've checked the official documentation and pub.dev documentation, and none of them mention a way to read the value of the family without changing it's value. I don't know if I'm using riverpod in a way I'm not supposed to or if families are not accessible outside of the build method, i'd appreciate it if you could tell me what I'm doing wrong here.

**Reply by TimWhiting** [reply]
testProvider(10) will only return 10 the first time you read it. Once you change the value (since it is a StateProvider) it will return whatever the current value of that provider is..

so this:
```dart
final val = ref.read(testProvider(10)) 
assert(val == 10);
ref.read(testProvider(10)).state += 10;
final val2 = ref.read(testProvider(10));
assert(val2 == 20);
```

```dart
final testProvider = StateProvider.family<int,int>((ref, testValue) { return testValue; })
```
This is still a provider creation function, (it gets called when the provider is created), and doesn't get called every time the provider is accessed. 

You can think of a family as a map that lazily creates entries the first time you access the key.
```dart
Map<int, StateProvider<int>> testProvider;
```

**Reply by imcouri** [reply]
seems like I completely misunderstood how families work, thank you for your help

**Reply by nero-angela** [reply]
The `10` passed in `ref.read(testProvider(10))` seems unnecessary. Is there any way not to pass it?

**Reply by rrousselGit** [reply]
No, it is absolutely necessary.

**Reply by wiztensai** [reply]
> No, it is absolutely necessary.

i think, this issue should be put to documentation.
